### PR TITLE
[SPARK-33428][SQL] Match the behavior of conv function to MySQL's

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?
Spark conv function is from MySQL and it's better to follow the MySQL behavior. MySQL returns the max unsigned long if the input string is too big, and Spark should follow it.

However, seems Spark has different behavior in two cases:

MySQL allows leading spaces but Spark does not.
If the input string is way too long, Spark fails with ArrayIndexOutOfBoundException

This patch now help conv follow behavior in those two cases
conv allows leading spaces
conv will return the max unsigned long when the input string is way too long 

### Why are the changes needed?
fixing it to match the behavior of conv function to the (almost) only one reference of another DBMS, MySQL

### Does this PR introduce _any_ user-facing change?
Yes, as pointed out above

### How was this patch tested?
Add test

### CODE REVIEW

**Comment by cloud-fan** [comment]
ok to test

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala [hunk 1]
```diff
@@ -203,6 +203,13 @@ class MathFunctionsSuite extends QueryTest with SharedSparkSession {
       df.selectExpr("""conv("9223372036854775807", 36, -16)"""), Row("-1")) // for overflow
   }
 
+  test("SPARK-33428 conv function has different behavior with mySQL's conv function") {
+    val df = Seq(("abc"), ("  abc"), ("abc  "), ("  abc  "),
+      ("aaaaaaa0aaaaaaa0aaaaaaa0aaaaaaa0aaaaaaa0aaaaaaa0aaaaaaa0aaaaaaa0aaaaaaa0")).toDF("num")
+    checkAnswer(df.select(conv('num, 16, 10)),
+      Seq(Row("2748"), Row("2748"), Row("2748"), Row("2748"), Row("18446744073709551615")))
+  }
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala:211]: Can you separate this into two test cases for two behaviour changes you described? Also, let's have a dedicated title for each test, e.g.) `SPARK-33428: conv should trim the input string`

**Comment by HyukjinKwon** [comment]
> Correctness

Can you fix the Pr description as well? I don't think it's fair to call it a correctness issue. I think we're fixing it to match the beahviour of `conv` function to the (almost) only one reference of another DBMS, MySQL.

File: sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala (hunk 1 above)
Comment (dgd-contributor) [reply sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala:215]: OK, I will

File: sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala [hunk 2]
```diff
@@ -203,6 +203,21 @@ class MathFunctionsSuite extends QueryTest with SharedSparkSession {
...
 
+  test("SPARK-33428 conv function should trim input string") {
+    val df = Seq(("abc"), ("  abc"), ("abc  "), ("  abc  ")).toDF("num")
+    checkAnswer(df.select(conv('num, 16, 10)),
+      Seq(Row("2748"), Row("2748"), Row("2748"), Row("2748")))
+    checkAnswer(df.select(conv('num, 16, -10)),
+      Seq(Row("2748"), Row("2748"), Row("2748"), Row("2748")))
+  }
+
+  test("SPARK-33428 conv function shouldn't raise error  input string is too big") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/MathFunctionsSuite.scala:214]:
```suggestion
  test("SPARK-33428 conv function shouldn't raise error input string is too big") {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberConverter.scala [hunk 3]
```diff
@@ -135,6 +138,7 @@ object NumberConverter {
     }
     decode(v, Math.abs(toBase), temp)
 
+
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberConverter.scala:141]:
```suggestion
```

**Review by HyukjinKwon — APPROVED** [review]
Looks good otherwise

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberConverter.scala (hunk 3 above)
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberConverter.scala]: thanks, fixed!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**